Feature/main files

bt-s opened the pull request:
Have a look at whether you approve of the changes that I have suggested. Also, if you think that the docstrings are incorrect, please leave a note why you think so and request for changes or change it yourself.

NOTE: all code regarding tiny ImageNet has been removed from this branch, for clarity's sake.
If we want to use this dataset, it can be added again later, and the relevant code can be found in the Git history.

bt-s committed:
Update .gitignore

bt-s committed:
WIP

bt-s committed:
Update gitignore for VIM

bt-s committed:
Create generic load_batch file

bt-s committed:
Write file introduction for RCNN_tf.py

bt-s committed:
Suppress warnings

bt-s committed:
Rename some variables and reformat code for improved readibility

bt-s committed:
Create flag for printing information about trainable parameters

bt-s committed:
WIP improve main function

bt-s committed:
Fix merge conflicts

bt-s committed:
Remove obsolete code from load_data.py

bt-s committed:
Provide docstrings following Google spec

leoTiez commented on RCNN_tf.py:
```diff
@@ -68,46 +93,81 @@ def rcl(
-            lambda x, recurrent_states, output: x < depth,
-            loop_body,
-            [0, cell_states, output],
+            cond=termination_cond(),
```
I don't think this works. `x`, `recurrent_states`, and `output` should be passed as parameter.
I will change that

bt-s replied:
I tested it and it seems to work just fine -- in my opinion it makes the code a bit easier to understand. The function termination_cond returns x, recurrent_states and output.

leoTiez commented on RCNN_tf.py:
```diff
@@ -21,25 +26,39 @@
+        filter_shape      (list): the shape of the 2D filter
+        num_of_data  (tf.Tensor): holds the batch size
+        processing_unit    (str): specifies whether a GPU or CPU is used
+        depth              (int): depth of the ???
```
I will add the description here

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+        filter_shape      (list): the shape of the 2D filter
+        stddev           (float): standard deviation
+        stride           (tuple): ???
+        padding            (str): ???
```
I can add the docstrings. For further information about stride and padding:
https://www.quora.com/What-are-the-roles-of-stride-and-padding-in-a-convolutional-neural-network

bt-s replied:
Ah, yeah: I know what these mean haha.
I just forgot to come back to it, my bad.

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+        input_data (tf.Tensor): 4D data tensor
+        pool_shape      (list): the shape of the 2D filter
+        stride         (tuple): ???
+        padding          (str): ???
```
Same here

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+            pool_stride_shape=[2, 2]
+        ):
+
+        # Input and output shape placeholders
```
That's not true. These are placeholders for the input and output data. However, the known dimensions should be set beforehand. This is why the shape is explicitly set. Since we want to be flexible with the number of inputs we can pass `None`

bt-s replied:
Good!

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
             self.input_placeholder,
-            self.output_placeholder,
-        )).batch(batch_size=self.num_data_placeholder)
+            self.output_placeholder,)
```
I will remove the comma here

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+        input_shape        (list): specification of the image input shape
+        output_shape       (list): specification of the class output shape
+        processing_unit     (str): specification of the PU to be used
+        learning_rate     (float): learning rate
```
Aren't the docstrings for the constructor normally within the constructor?

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
             rcl_layer_1 = rcl(
                 input_data=pooling_1,
-                num_input_channels=num_filter,
+                num_input_chans=num_filter,
```
Fine for now, but in future I would like to avoid renaming variables, if there is no reason for that. It is totally appropriate when they are deprecated, misleading, or not descriptive.
This, however, adds points of failure and blows up the changes on github which are needed to be reviewed.

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+                num_input_chans=num_filter,
                 num_of_data=self.num_data_placeholder,
-                device_name=self.device_name,
-                filter_shape=rec_conv_layer_filter_shape,
```
Same with this one

leoTiez commented on RCNN_tf.py:
```diff
@@ -396,93 +548,168 @@ def train(
+                    self.input_placeholder: val_data_feats,
                     self.output_placeholder: val_data_labels,
-                    self.num_data_placeholder: val_data_features.shape[0]
+                    self.num_data_placeholder: val_data_feats.shape[0]
```
Again: Fine for now, but this can be very dangerous and break things.

bt-s replied:
I did a thorough test, so don't worry.

leoTiez commented on RCNN_tf.py:
```diff
@@ -396,93 +548,168 @@ def train(
+
+    else:
+        raise Exception("dataset has to be one of 'mnist', 'cifar10', "
+                        "'cifar100' or 'tinynet'.")
```
I will remove the tinynet

bt-s replied:
Well spotted!

leoTiez commented on RCNN_tf.py:
```diff
@@ -396,93 +548,168 @@ def train(
+        #batch_size=100,
+        #epochs=epochs_,
+        #create_graph=True,
+        #print_vars=True
```
Is there a reason why they are commented out?

bt-s replied:
Yes: I just wanted to signal that these are the standard input parameters.
They could be changed when we are tweaking the model.

leoTiez commented on RCNN_tf.py:
```diff
@@ -10,9 +10,14 @@
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2' # suppress warnings
+
 import tensorflow as tf
+tf.logging.set_verbosity(tf.logging.ERROR) # suppress warnings
```
Are both lines
`os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'``and
`tf.logging.set_verbosity(tf.logging.ERROR) `
really necessary to suppress the warnings or is it more duplicating

bt-s replied:
They are both necessary, since one suppresses warnings and the other one suppress the "I, Tensorflow did .... bla bla bla" output.

leoTiez commented on RCNN_tf.py:
```diff
@@ -116,209 +176,292 @@ def loop_body(x, recurrent_states, output):
+            self.output_placeholder)
+            ).shuffle(buffer_size=shuf_buf_size
+            ).repeat(
+            ).batch(batch_size=self.num_data_placeholder)
```
This line splitting is very hard to read I find

bt-s replied:
Yeah, I agree: change it if you have a better formatting in mind.

In Javascript I would write something like:

```
self.output_placeholder))
     .shuffle(buffer_size=shuf_buf_size)
     .repeat()
     .batch(batch_size=self.num_data_placeholder)
```

Unfortunately Python cannot parse that....

leoTiez commented:
I will add `Pillow` to the requirements.txt. Please add every library to the requirements.txt, if it is not a standard library

leoTiez commented:
But I see right now you do not even use it. Alright, then I will remove these lines

leoTiez committed:
Udate: Removed unnecessary library imports

leoTiez committed:
Update: added missing docstrings

Also. tuple to list for stride parameter to be consistent.

While loop in cpu execution has been slightly adapted
since empirical tests revealed it works better.
Todo comment left for investigation

leoTiez committed:
Merge pull request #25 from leoTiez/bugfix/accuracy

Bugfix/accuracy

leoTiez commented:
All the issues have been discussed. Ready for merge
